### PR TITLE
gh-114572: Fix locking in cert_store_stats and get_ca_certs

### PR DESCRIPTION
**Pull request by davidben**
cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with X509_STORE_get0_objects, but reading the result requires a lock. See https://github.com/openssl/openssl/pull/23224 for details.

Instead, use X509_STORE_get1_objects, newly added in that PR. X509_STORE_get1_objects does not exist in current OpenSSLs, but we can polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Issue: gh-114572

### CODE REVIEW

File: Modules/_ssl.c
Comment (sethmlarson) [inline Modules/_ssl.c:4684]: Do we need to add a null check here as well?

File: Modules/_ssl.c
Comment (sethmlarson) [inline Modules/_ssl.c:4586]: Noting this is a faithful backport of https://github.com/openssl/openssl/commit/08cecb4448e990f7914ec1df97b1ee0ca9031643

File: Modules/_ssl.c
Comment (sethmlarson) [inline Modules/_ssl.c:4557]: From looking at all the OpenSSL docs this appears to be correct? I'm less versed with the OpenSSL API than the submitter :)

**Review by sethmlarson — COMMENTED** [review]
Thanks for submitting this David!

This change feels newsworthy, can we get a news entry for this change using [blurb](https://devguide.python.org/core-developers/committing/#how-to-add-a-news-entry)?

I've reviewed the approach and I'm happy with it but will need a core developers' review before it can be merged.

**Comment by sethmlarson** [comment]
Also noting that this PR can't be directly backported, some of the functions it relies on are only in OpenSSL 1.1.0+ (specifically the locking functions).

File: Modules/_ssl.c
Comment (davidben) [inline Modules/_ssl.c]: Note to self: add some casts here. Looks like OpenSSL only made these functions const-correct in 3.0, which is why we're getting warnings in some places. Pfft.

Comment (davidben) [reply Modules/_ssl.c]: Done.

File: Modules/_ssl.c
Comment (davidben) [reply Modules/_ssl.c:4684]: Ah whoops. Done.

**Comment by davidben** [comment]
> This change feels newsworthy, can we get a news entry for this change using [blurb](https://devguide.python.org/core-developers/committing/#how-to-add-a-news-entry)?

Done.

**Comment by davidben** [comment]
Anything left to do on my end on this one?

**Review by alex — APPROVED** [review]

**Comment by alex** [comment]
@sethmlarson This LGTM, wasn't sure if there was anything you wanted to verify before merging.

**Comment by sethmlarson** [comment]
@alex I think this is good to go! Thank you :)

**Comment by alex** [comment]
This doesn't look connected to this PR?

**Comment by sethmlarson** [comment]
@alex I don't think it is, thanks for digging in too :)

**Pull request merged**
